packages/functions/src/resample.ts

```diff
@@ -70,92 +121,21 @@ export function resample(_options: ResampleOptions = RESAMPLE_DEFAULTS): Transfo
-	const input = sampler.getInput()!.clone().setSparse(false);
-	const output = sampler.getOutput()!.clone().setSparse(false);
-
-	const tolerance = options.tolerance as number;
-	const interpolation = sampler.getInterpolation();
-
-	const lastIndex = input.getCount() - 1;
-	const tmp: number[] = [];
-	const value: number[] = [];
-	const valueNext: number[] = [];
-	const valuePrev: number[] = [];
-
-	let writeIndex = 1;
-
-	for (let i = 1; i < lastIndex; ++i) {
-		const timePrev = input.getScalar(writeIndex - 1);
-		const time = input.getScalar(i);
-		const timeNext = input.getScalar(i + 1);
-		const t = (time - timePrev) / (timeNext - timePrev);
-
-		let keep = false;
-
-		// Remove unnecessary adjacent keyframes.
-		if (time !== timeNext && (i !== 1 || time !== input.getScalar(0))) {
-			output.getElement(writeIndex - 1, valuePrev);
-			output.getElement(i, value);
-			output.getElement(i + 1, valueNext);
-
-			if (interpolation === 'LINEAR' && path === 'rotation') {
-				// Prune keyframes colinear with prev/next keyframes.
-				const sample = slerp(tmp as quat, valuePrev as quat, valueNext as quat, t) as number[];
-				const angle = getAngle(valuePrev as quat, value as quat) + getAngle(value as quat, valueNext as quat);
-				keep = !MathUtils.eq(value, sample, tolerance) || angle + Number.EPSILON >= Math.PI;
-			} else if (interpolation === 'LINEAR') {
-				// Prune keyframes colinear with prev/next keyframes.
-				const sample = vlerp(tmp, valuePrev, valueNext, t);
-				keep = !MathUtils.eq(value, sample, tolerance);
-			} else if (interpolation === 'STEP') {
-				// Prune keyframes identical to prev/next keyframes.
-				keep = !MathUtils.eq(value, valuePrev) || !MathUtils.eq(value, valueNext);
-			}
-		}
-
-		// In-place compaction.
-		if (keep) {
-			if (i !== writeIndex) {
-				input.setScalar(writeIndex, input.getScalar(i));
-				output.setElement(writeIndex, output.getElement(i, tmp));
-			}
-			writeIndex++;
-		}
-	}
-
-	// Flush last keyframe (compaction looks ahead).
-	if (lastIndex > 0) {
-		input.setScalar(writeIndex, input.getScalar(lastIndex));
-		output.setElement(writeIndex, output.getElement(lastIndex, tmp));
-		writeIndex++;
-	}
+function toFloat32Array(
+	srcArray: TypedArray,
+	componentType: GLTF.AccessorComponentType,
+	normalized: boolean
+): Float32Array {
+	const dstArray = new Float32Array(srcArray);
+	if (!normalized) return dstArray;
 
-	// If the sampler was optimized, truncate and save the results. If not, clean up.
-	if (writeIndex !== input.getCount()) {
-		input.setArray(input.getArray()!.slice(0, writeIndex));
-		output.setArray(output.getArray()!.slice(0, writeIndex * output.getElementSize()));
-		sampler.setInput(input);
-		sampler.setOutput(output);
-	} else {
-		input.dispose();
-		output.dispose();
+	for (let i = 0; i < dstArray.length; i++) {
+		dstArray[i] = MathUtils.decodeNormalizedInt(dstArray[i], componentType);
```

Comment:
Not benchmarked yet, but since denormalize is a pure algorithm, it might worth to implement it in wasm, sample [here](https://github.com/kzhsw/keyframe-resample-c/blob/81b04da50ba05128ac1b2a4fe0043d603a31178b/normalize.c#L481).

Reply:
Maybe worth considering in the https://github.com/donmccurdy/keyframe-resample-wasm module — but I think I'm satisfied enough, as-is. I don't think I've ever seen a file using normalized accessors for animation, and it seems too early to add complexity to potentially optimize that.